include/odp/api/spec/comp.h

```diff
@@ -0,0 +1,844 @@
+ * odp_comp_result() on 1st failed packet (i.e. n+1th packet where
+ * n=number of packets returned) to ensure if error code is
+ * ODP_COMP_ERR_OUT_OF_SPACE.
+ *
+ * In case of ODP_COMP_ERR_OUT_OF_SPACE, application should keep on calling
+ * odp_comp_op() API with more output buffer until call returns
+ * with num_pkt or with result as ODP_COMP_ERR_NONE or any other
+ * failure code.
+ *
+ * @param         pkt_in   Packets to be processed
+ * @param[in,out] pkt_out  Packet handle array specifying resulting packets
+ * @param         param    Operation parameters array
+ * @param         num_pkt  Number of packets to be processed
+ *
+ * @return Number of input packets consumed (0 ... num_pkt)
+ * @retval <0 on failure
+ */
+int odp_comp_decompress(const odp_packet_t pkt_in[],
+			odp_packet_t pkt_out[],
+			odp_comp_packet_op_param_t param[],
+			int num_pkt);
+
+/**
+ * Comp packet operation asynchronous
+ *
+ * Performs the ASYNC compression operations on the packets.
+ * If session is created in ODP_COMP_ASYNC mode, event will be queued
+ * to completion queue. Application should monitor ODP_EVENT_PACKET with
+ * subtype ODP_EVENT_PACKET_COMP on queue.
+ *
+ * If session is created in ODP_COMP_SYNC mode, behavior is undefined.
+ *
+ * Caller should initialize pkt_out with valid output
+ * packet handles.All arrays should be of num_pkt size. Resulting packets
+ * are returned through events.
+ *
+ * In case of partially processed array i.e. when
+ * number of packets returned < num_pkt, application should call
+ * odp_comp_result() on 1st failed packet (i.e. n+1th packet where
+ * n=number of packets returned) to ensure if error code is
+ * ODP_COMP_ERR_OUT_OF_SPACE.
+ *
+ * In case of ODP_COMP_ERR_OUT_OF_SPACE, application should keep on calling
+ * odp_comp_op() API with more output buffer until call returns
+ * with num_pkt or with result as ODP_COMP_ERR_NONE or any other
+ * failure code.
+ *
+ * For compression + hash, call returns with hash appended to the end of
+ * last processed chunk of data.
+ *
+ * User should compute processed data len = total output len - digest_len, where
+ * digest_len queried through odp_comp_hash_alg_capability().
+ *
+ * Please note it is always recommended that application using async mode,
+ * provide sufficiently large buffer size to avoid ODP_COMP_ERR_OUT_OF_SPACE.
+ * Else it is recommended that application maintain relevant context
+ * with respect to each input processing request to correctly identify
+ * its corresponding enqueued event.
+ *
+ *
+ * @param pkt_in   Packets to be processed
+ * @param pkt_out  Packet handle array specifying resulting packets
+ * @param param    Operation parameters array
+ * @param num_pkt  Number of packets to be processed
+ *
+ * @return Number of input packets consumed (0 ... num_pkt)
+ * @retval <0 on failure
+ */
+int odp_comp_compress_enq(const odp_packet_t pkt_in[],
+			  odp_packet_t pkt_out[],
+			  odp_comp_packet_op_param_t param[],
```

Comment:
Ideally both pkt_out and param should be constants here. One should not return information through them.

Reply:
user will pass me output packet handle, which I will return via event. we can keep packet array as constant but for params I will have to think and come back on this for async behavior

Reply:
As I wrote, I'd suggest to return output length via result. As a short cut, you might want to implement default behaviour, when compression routines will just truncate output packet to the proper length.

Reply:
For async its fine, could you confirm on Sync side -- Do I need to call result API ?